[CALCITE-2742] Update RexImpTable to use DataContext to retrieve USER and SYSTEM_USER

siddharthteotia opened the pull request.

siddharthteotia commented:
ping?

cc @laurentgo , @vladimirdolzhenko 

I am not sure exactly who can/should review this piece of code. Can I please get this changeset reviewed? Please add others if someone else should be looking at this code.

risdenk commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -43,6 +43,7 @@
 import org.apache.calcite.util.TimeWithTimeZoneString;
 import org.apache.calcite.util.TimestampWithTimeZoneString;
 
+import com.google.common.base.Optional;
```
Do we need to use Guava here? Java 8 optional should work?

siddharthteotia replied:
Thanks @risdenk. We can use Java 8. I have made the change.

vlsi commented on core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java:
```diff
@@ -175,6 +176,58 @@ private void checkConstant(final Object operand,
+            }
+          });
+        } catch (Exception e) {
+          throw new RuntimeException(e);
```
Just a minor note: `throw org.apache.calcite.util.TestUtil#rethrow(java.lang.Throwable)` should be used to rethrow exceptions in the test code.

siddharthteotia replied:
Thanks @vlsi. I have made the change.

julianhyde replied:
@vlsi Is it possible to add a forbidden-apis check to flag these violations?

vlsi replied:
@julianhyde , I've no idea. It could probably work if we split current execution in two:

```
<execution>
            <goals>
              <goal>check</goal>
              <goal>testCheck</goal>
            </goals>
</execution>
```

However I find replacing Apache JMeter's Ant-based build with a Gradle-based one to be a more fascinating thing a the moment.

PS. The more interesting thing here is `new Function` which should be expressed as a lambda to reduce verbosity. 

PPS. I do like the way Gradle feels. It just automatically rebuilds the required modules. For instance, in Maven you need to "mvn install" the core before you run "csv tests", however Gradle does that transparently, and it can even do that across projects. It is really amazing.

vlsi commented:
LGTM

julianhyde commented on core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java:
```diff
@@ -175,6 +177,58 @@ private void checkConstant(final Object operand,
+    long val = System.currentTimeMillis() / 1000 * 1000;
+    testContextLiteral(SqlStdOperatorTable.CURRENT_TIMESTAMP,
+                       DataContext.Variable.CURRENT_TIMESTAMP, val);
+  }
```
I think this test will fail intermittently. Rounding to seconds will make it less frequent but doesn't eliminate the problem.

siddharthteotia replied:
How can we address this?

julianhyde replied:
Well, if the test starts at 10,000 and finishes at 12,000 then any timestamp value between 10,000 and 12,000 should be considered valid.

julianhyde commented on core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java:
```diff
@@ -175,6 +177,58 @@ private void checkConstant(final Object operand,
+  }
+
+  /**
+   * Ensure that for a given context operator, the correct value is retrieved from the DataContext.
```
"Ensures" not "Ensure".

siddharthteotia replied:
done.

julianhyde commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -1966,6 +1967,20 @@ public static TimeZone timeZone(DataContext root) {
   }
 
+  /** SQL {@code USER} function. */
+  @NonDeterministic
```
why is this non-deterministic?

siddharthteotia replied:
Thanks. Yes this should be deterministic. It was done to be consistent with other functions in the class

julianhyde commented on core/src/test/java/org/apache/calcite/rex/RexExecutorTest.java:
```diff
@@ -175,6 +177,58 @@ private void checkConstant(final Object operand,
+  @Test public void testUserFromContext() throws Exception {
+    testContextLiteral(SqlStdOperatorTable.USER,
+                       DataContext.Variable.USER, "happyCalciteUser");
+  }
```
coding style is to indent arguments 4 spaces, not to line up with '('.

siddharthteotia replied:
done.

asereda-gs commented on core/src/test/java/org/apache/calcite/test/JdbcTest.java:
```diff
@@ -2385,17 +2385,17 @@ private void checkNullableTimestamp(CalciteAssert.Config config) {
     CalciteAssert.hr()
         .query(
-            "select substring(\"name\", \"deptno\"+case when user <> 'sa' then 1 end) from \"hr\".\"emps\"")
+            "select substring(\"name\", \"deptno\"+case when CURRENT_PATH <> '' then 1 end) from \"hr\".\"emps\"")
```
Is it possible to set `user` to `sa` for current test instead of changing the query ?

asereda-gs commented:
Hi @siddharthteotia can you please squash commits into one and fix existing conflicts ? Add your full name in parenthesis in commit message.  
Also please address Julian comments about flaky behavior in `testTimestampFromContext`.

asereda-gs commented on core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:
```diff
@@ -1966,6 +1967,20 @@ public static TimeZone timeZone(DataContext root) {
+  /** SQL {@code USER} function. */
+  @Deterministic
+  public static String user(DataContext root) {
+    return java.util.Optional.ofNullable((String) DataContext.Variable.USER.get(root))
```
Pls remove `java.util` package. `java.util.Optional` is already imported.

siddharthteotia replied:
done.

siddharthteotia commented:
@asereda-gs , @julianhyde , I have addressed the review comments, rebased and squashed the commits. Please review when you get a chance.

siddharthteotia commented:
ping @julianhyde , @asereda-gs , @vlsi

siddharthteotia commented:
@julianhyde , @asereda-gs , @vlsi , can one of you please review this?

asereda-gs commented:
@siddharthteotia  did you address Julian's comment about flaky test (see `testTimestampFromContext`) eg. check for a range of correct values ?

Also why `JdbcTest.testReuseExpressionWhenNullChecking3` was changed from `user = 'sa' ` to `CURRENT_PATH = ''`. There is no way to set that variable explicitly in test ?

siddharthteotia commented:
@asereda-gs , yes I have addressed the review comments. W.r.t change from user = 'sa', I couldn't find another way of setting that in the test. Please review the changes.

siddharthteotia commented:
rebased.

julianhyde commented:
Reviewing now. Please see comments in the JIRA case.

siddharthteotia commented:
> Reviewing now. Please see comments in the JIRA case.

Thanks for reviewing. The build is passing. We can merge this along with your revision or I can cherry-pick your commit and update this PR

siddharthteotia commented:
ping @julianhyde

siddharthteotia commented:
> > Reviewing now. Please see comments in the JIRA case.
> 
> Thanks for reviewing. The build is passing. We can merge this along with your revision or I can cherry-pick your commit and update this PR

@julianhyde , I have cherry-picked your revision and updated this PR. I think we can merge this once the build finishes.
